app/src/src_nonfreenet/org/breezyweather/sources/ekuk/EkukService.kt

```diff
@@ -0,0 +1,180 @@
+        val today = formatter.format(now.time)
+        now.add(Calendar.DATE, -1)
+        val yesterday = formatter.format(now.time)
+
+        val failedFeatures = mutableListOf<SourceFeature>()
+        val airQuality = if (SourceFeature.AIR_QUALITY in requestedFeatures) {
+            if (!airQualityStation.isNullOrEmpty()) {
+                mApi.getObservations(
+                    station = airQualityStation,
+                    indicators = EKUK_AIR_QUALITY_INDICATORS,
+                    range = "$yesterday,$today"
+                ).onErrorResumeNext {
+                    failedFeatures.add(SourceFeature.AIR_QUALITY)
+                    Observable.just(emptyList())
+                }
+            } else {
+                // Do not fail: airQualityStation is empty if the nearest one is > 50km away
+                Observable.just(emptyList())
+            }
+        } else {
+            Observable.just(emptyList())
+        }
+
+        val pollen = if (SourceFeature.POLLEN in requestedFeatures) {
+            if (!pollenStation.isNullOrEmpty()) {
+                mApi.getObservations(
+                    station = pollenStation,
+                    indicators = EKUK_POLLEN_INDICATORS,
+                    range = "$yesterday,$today"
+                ).onErrorResumeNext {
+                    failedFeatures.add(SourceFeature.POLLEN)
+                    Observable.just(emptyList())
+                }
+            } else {
+                // Do not fail: pollenStation is empty if the nearest one is > 50km away
+                Observable.just(emptyList())
+            }
+        } else {
+            Observable.just(emptyList())
+        }
+
+        return Observable.zip(airQuality, pollen) {
+                airQualityResult: List<EkukObservationsResult>,
+                pollenResult: List<EkukObservationsResult>,
+            ->
+            convert(
+                airQualityResult = if (SourceFeature.AIR_QUALITY in requestedFeatures) {
+                    airQualityResult
+                } else {
+                    null
+                },
+                pollenResult = if (SourceFeature.POLLEN in requestedFeatures) {
+                    pollenResult
+                } else {
+                    null
+                },
+                failedFeatures = failedFeatures
+            )
+        }
+    }
+
+    override fun needsLocationParametersRefresh(
+        location: Location,
+        coordinatesChanged: Boolean,
+        features: List<SourceFeature>,
+    ): Boolean {
+        if (coordinatesChanged) return true
+        val airQualityStation = location.parameters.getOrElse(id) { null }?.getOrElse("airQualityStation") { null }
+        val pollenStation = location.parameters.getOrElse(id) { null }?.getOrElse("pollenStation") { null }
+
+        // Do not check for empty, because a valid station should be within 50km of selected location
+        return (airQualityStation == null) || (pollenStation == null)
```

Comment:
You only need an `airQualityStation` if `SourceFeature.AIR_QUALITY in features` and a `pollenStation` if `SourceFeature.POLLEN in features`

Reply:
Fixed